cpp/include/cuspatial/experimental/detail/point_polygon_distance.cuh

```diff
@@ -0,0 +1,170 @@
+#include <cuspatial/detail/utility/linestring.cuh>
+#include <cuspatial/detail/utility/offset_to_keys.cuh>
+#include <cuspatial/detail/utility/zero_data.cuh>
+#include <cuspatial/error.hpp>
+#include <cuspatial/experimental/detail/algorithm/is_point_in_polygon.cuh>
+#include <cuspatial/experimental/ranges/range.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/tabulate.h>
+
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief For each point in the multipoint, compute point-in-multipolygon in corresponding pair.
+ */
+template <typename MultiPointRange, typename MultiPolygonRange>
+struct point_in_multipolygon_test_functor {
+  MultiPointRange multipoints;
+  MultiPolygonRange multipolygons;
+
+  point_in_multipolygon_test_functor(MultiPointRange multipoints, MultiPolygonRange multipolygons)
+    : multipoints(multipoints), multipolygons(multipolygons)
+  {
+  }
+
+  template <typename IndexType>
+  uint8_t __device__ operator()(IndexType pidx)
+  {
+    auto point        = thrust::raw_reference_cast(multipoints.point(pidx));
+    auto geometry_idx = multipoints.geometry_idx_from_point_idx(pidx);
+
+    bool intersects = false;
+    for (auto polygon : multipolygons[geometry_idx]) {
+      intersects = intersects || is_point_in_polygon(point, polygon);
+    }
+    return static_cast<uint8_t>(intersects);
+  }
+};
+
+/**
+ * @brief Kernel to compute the distance between pairs of point and polygon.
+ */
+template <class MultiPointRange,
+          class MultiPolygonRange,
+          class IntersectionRange,
+          class OutputIterator>
+void __global__ pairwise_point_polygon_distance_kernel(MultiPointRange multipoints,
+                                                       MultiPolygonRange multipolygons,
+                                                       IntersectionRange intersects,
+                                                       OutputIterator distances)
+{
+  using T = typename MultiPointRange::element_t;
+
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multipolygons.num_points();
+       idx += gridDim.x * blockDim.x) {
+    auto geometry_idx = multipolygons.geometry_idx_from_segment_idx(idx);
+    if (geometry_idx == MultiPolygonRange::INVALID_INDEX) continue;
+
+    if (intersects[geometry_idx]) {
+      // Leading thread of the pair writes to the output
+      if (multipolygons.is_first_point_of_multipolygon(idx, geometry_idx))
+        distances[geometry_idx] = T{0.0};
+      continue;
+    }
+
+    T dist_squared = std::numeric_limits<T>::max();
+    auto [a, b]    = multipolygons.get_segment(idx);
+    for (vec_2d<T> point : multipoints[geometry_idx]) {
```

Comment:
`thrust::reduce(thrust::seq, ...)`? Not sure about performance.

Reply:
Needs benchmarking, but better redability.

Reply:
Agree. In this case, while a `reduce` signifies intent better, the loop is already pretty clear.